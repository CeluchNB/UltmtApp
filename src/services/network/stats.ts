import { API_KEY, API_URL_V1 } from '@env'
import axios, { AxiosResponse } from 'axios'

export const getPlayerStats = async (id: string): Promise<AxiosResponse> => {
    return await axios.get(`${API_URL_V1}/stats/player/${id}`, {
        headers: {
            'X-API-Key': API_KEY,
        },
    })
}

export const filterPlayerStats = async (
    id: string,
    teams: string[],
    games: string[],
): Promise<AxiosResponse> => {
    let url = `${API_URL_V1}/stats/filter/player/${id}`
    if (teams.length > 0) {
        url += `?teams=${teams.join(',')}`
    }
    if (games.length > 0) {
        if (teams.length > 0) {
            url += `&games=${games.join(',')}`
        } else {
            url += `?games=${games.join(',')}`
        }
    }
    return await axios.get(url, {
        headers: {
            'X-API-Key': API_KEY,
        },
    })
}

export const getGameStats = async (id: string): Promise<AxiosResponse> => {
    return await axios.get(`${API_URL_V1}/stats/game/${id}`, {
        headers: { 'X-API-Key': API_KEY },
    })
}

export const getGameStatsByTeam = async (
    gameId: string,
    teamId: string,
): Promise<AxiosResponse> => {
    return await axios.get(
        `${API_URL_V1}/stats/filter/game/${gameId}?team=${teamId}`,
        { headers: { 'X-API-Key': API_KEY } },
    )
}

export const getTeamStats = async (id: string): Promise<AxiosResponse> => {
    return await axios.get(`${API_URL_V1}/stats/team/${id}`, {
        headers: { 'X-API-Key': API_KEY },
    })
}

export const getTeamStatsByGame = async (
    teamId: string,
    gameIds: string[],
): Promise<AxiosResponse> => {
    let url = `${API_URL_V1}/stats/filter/team/${teamId}`
    if (gameIds.length > 0) {
        url += `?games=${gameIds.join(',')}`
    }
    return await axios.get(url, { headers: { 'X-API-Key': API_KEY } })
}

export const getConnectionStats = async (
    throwerId: string,
    receiverId: string,
): Promise<AxiosResponse> => {
    return await axios.get(
        `${API_URL_V1}/stats/connection?thrower=${throwerId}&receiver=${receiverId}`,
        {
            headers: { 'X-API-Key': API_KEY },
        },
    )
}

export const filterConnectionStats = async (
    throwerId: string,
    receiverId: string,
    games: string[],
    teams: string[],
): Promise<AxiosResponse> => {
    let url = `${API_URL_V1}/stats/filter/connection?thrower=${throwerId}&receiver=${receiverId}`
    if (teams.length > 0) {
        url += `&teams=${teams.join(',')}`
    }
    if (games.length > 0) {
        url += `&games=${games.join(',')}`
    }
    return await axios.get(url, {
        headers: {
            'X-API-Key': API_KEY,
        },
    })
}
