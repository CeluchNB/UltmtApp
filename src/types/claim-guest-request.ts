import { DisplayTeam } from './team'
import { DisplayUser } from './user'

export interface ClaimGuestRequest {
    _id: string
    guestId: string
    userId: string
    teamId: string
    guest: DisplayUser
    user: DisplayUser
    team: DisplayTeam
    status: string
}
