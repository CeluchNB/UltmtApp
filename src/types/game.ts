import { Tournament } from './tournament'
import { DisplayTeam, GuestTeam } from './team'
import { DisplayUser, GuestUser } from './user'
export interface DisplayGame {
    opponent: DisplayTeam
    teamScore: number
    opponentScore: number
}

export interface CreateGame {
    creator: DisplayUser
    teamOne: DisplayTeam
    teamTwo: GuestTeam
    teamTwoDefined: boolean
    scoreLimit: number
    halfScore: number
    startTime: Date
    softcapMins: number
    hardcapMins: number
    playersPerPoint: number
    timeoutPerHalf: number
    floaterTimeout: boolean
    tournament?: Tournament
}

export interface Game {
    _id: string
    creator: DisplayUser
    teamOne: DisplayTeam
    teamTwo: GuestTeam
    teamTwoDefined: boolean
    scoreLimit: number
    halfScore: number
    startTime: Date
    softcapMins: number
    hardcapMins: number
    playersPerPoint: number
    timeoutPerHalf: number
    floaterTimeout: boolean
    tournament?: Tournament
    teamOneScore: number
    teamTwoScore: number
    teamOneActive: boolean
    teamTwoActive: boolean
    teamOnePlayers: DisplayUser[]
    teamTwoPlayers: DisplayUser[]
    resolveCode: string
    points: string[]
}

export interface SearchParams {
    q?: string
    live?: boolean
    after?: Date
    before?: Date
}
