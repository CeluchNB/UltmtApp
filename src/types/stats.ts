import { DisplayUser } from './user'
export interface DisplayStat {
    name: string
    value: number
    points: number
}

export interface PlayerStats {
    goals: number
    assists: number
    blocks: number
    throwaways: number
    drops: number
    stalls: number
    touches: number
    catches: number
    completedPasses: number
    droppedPasses: number
    callahans: number
    pointsPlayed: number
    pulls: number
    wins: number
    losses: number
}

export interface CalculatedPlayerStats {
    plusMinus: number
    catchingPercentage: number
    throwingPercentage: number
    ppGoals: number
    ppAssists: number
    ppThrowaways: number
    ppDrops: number
    ppBlocks: number
    winPercentage: number
}

export interface AllPlayerStats extends PlayerStats, CalculatedPlayerStats {}

export interface IdentifiedPlayerStats extends DisplayUser, AllPlayerStats {
    games: string[]
    teams: string[]
}
