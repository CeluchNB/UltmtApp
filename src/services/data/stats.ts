import * as Constants from '../../utils/constants'
import { getInitialPlayerData } from '../../../fixtures/utils'
import { throwApiError } from '../../utils/service-utils'
import {
    AllPlayerStats,
    FilteredGameStats,
    FilteredTeamStats,
    GameStats,
    IdentifiedPlayerStats,
    PlayerStats,
} from '../../types/stats'
import { addPlayerStats, calculatePlayerStats } from '../../utils/stats'
import {
    filterPlayerStats as networkFilterPlayerStats,
    getGameStats as networkGetGameStats,
    getGameStatsByTeam as networkGetGameStatsByTeam,
    getPlayerStats as networkGetPlayerStats,
    getTeamStats as networkGetTeamStats,
    getTeamStatsByGame as networkGetTeamStatsByGame,
} from '../network/stats'

export const getPlayerStats = async (
    id: string,
): Promise<IdentifiedPlayerStats> => {
    try {
        const response = await networkGetPlayerStats(id)
        const { player } = response.data
        return player
    } catch (e) {
        return throwApiError(e, Constants.UNABLE_TO_GET_PLAYER_STATS)
    }
}

export const filterPlayerStats = async (
    id: string,
    teams: string[],
    games: string[],
): Promise<AllPlayerStats> => {
    try {
        const response = await networkFilterPlayerStats(id, teams, games)
        const { stats } = response.data

        const statAggregate: PlayerStats = stats.reduce(
            (a: PlayerStats, b: PlayerStats) => addPlayerStats(a, b),
            getInitialPlayerData({}),
        )
        return calculatePlayerStats(statAggregate)
    } catch (e) {
        return throwApiError(e, Constants.UNABLE_TO_GET_PLAYER_STATS)
    }
}

export const getGameStats = async (id: string): Promise<GameStats> => {
    try {
        const response = await networkGetGameStats(id)
        const { game } = response.data
        return game
    } catch (e) {
        return throwApiError(e, Constants.UNABLE_TO_GET_GAME_STATS)
    }
}

export const getGameStatsByTeam = async (
    gameId: string,
    teamId: string,
): Promise<FilteredGameStats> => {
    try {
        const response = await networkGetGameStatsByTeam(gameId, teamId)
        const { game } = response.data

        return game
    } catch (e) {
        return throwApiError(e, Constants.UNABLE_TO_GET_GAME_STATS)
    }
}

export const getTeamStats = async (
    teamId: string,
): Promise<FilteredTeamStats> => {
    try {
        const response = await networkGetTeamStats(teamId)
        const { team } = response.data

        return team
    } catch (e) {
        return throwApiError(e, Constants.UNABLE_TO_GET_TEAM_STATS)
    }
}

export const getTeamStatsByGame = async (
    teamId: string,
    gameIds: string[],
): Promise<FilteredTeamStats> => {
    try {
        const response = await networkGetTeamStatsByGame(teamId, gameIds)
        const { team } = response.data

        return team
    } catch (e) {
        return throwApiError(e, Constants.UNABLE_TO_GET_TEAM_STATS)
    }
}
