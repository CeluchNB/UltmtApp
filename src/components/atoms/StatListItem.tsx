import * as React from 'react'
import { DisplayStat } from '../../types/stats'
import { useTheme } from '../../hooks'
import { StyleSheet, Text, View } from 'react-native'

interface StatListItemProps {
    stat: DisplayStat
}

const StatListItem: React.FC<StatListItemProps> = ({ stat }) => {
    const {
        theme: { colors, size, weight },
    } = useTheme()

    const styles = StyleSheet.create({
        container: {
            minWidth: 150,
        },
        statName: {
            color: colors.gray,
            fontSize: size.fontTwenty,
            fontWeight: weight.bold,
        },
        value: {
            color: colors.textPrimary,
            fontSize: size.fontThirty,
            fontWeight: weight.bold,
        },
        ratioView: {
            backgroundColor: colors.textSecondary,
            width: 100 * Math.min(1, stat.value / stat.points),
            height: 10,
            borderRadius: 5,
        },
    })

    return (
        <View style={styles.container}>
            <Text style={styles.statName}>{stat.name}</Text>
            <Text style={styles.value}>{stat.value}</Text>
            <View style={styles.ratioView} />
        </View>
    )
}

export default StatListItem
