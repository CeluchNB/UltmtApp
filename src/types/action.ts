import { DisplayUser } from './user'
import { getUserDisplayName } from '../utils/player'
import { DisplayTeam, TeamNumber } from './team'
import { isRetainingPossession, isScore, isTurnover } from '../utils/action'

export interface Action {
    action: ServerAction
    reporterDisplay: string
    viewerDisplay: string
    setTags(tags: string[]): void
    setPlayersAndUpdateViewerDisplay(
        playerOne?: DisplayUser,
        playerTwo?: DisplayUser,
    ): void
}

export enum ActionType {
    PULL = 'Pull',
    CATCH = 'Catch',
    DROP = 'Drop',
    THROWAWAY = 'Throwaway',
    BLOCK = 'Block',
    PICKUP = 'Pickup',
    TEAM_ONE_SCORE = 'TeamOneScore',
    TEAM_TWO_SCORE = 'TeamTwoScore',
    TIMEOUT = 'Timeout',
    SUBSTITUTION = 'Substitution',
    CALL_ON_FIELD = 'CallOnField',
}

export type SubscriptionType = 'client' | 'undo' | 'error' | 'point'
export type SubscriptionObject = {
    [x in SubscriptionType]: (data: any) => Promise<void> | void
}

export interface ClientAction {
    actionType: ActionType
    playerOne?: DisplayUser
    playerTwo?: DisplayUser
    tags: string[]
}

export interface Comment {
    user: DisplayUser
    comment: string
    commentNumber: number
}

export interface ServerAction extends ClientAction {
    comments: Comment[]
    actionNumber: number
}

export interface LiveServerAction extends ServerAction {
    teamNumber: TeamNumber
}

export interface SavedServerAction extends ServerAction {
    _id: string
    team: DisplayTeam
}

export interface ActionListData {
    actionList: Action[]
}

export class PlayerActionListData implements ActionListData {
    actionList: Action[]
    constructor(
        playerOne: DisplayUser,
        actionStack: LiveServerAction[],
        team: TeamNumber,
        pulling: boolean,
    ) {
        this.actionList = getPlayerActionList(
            playerOne,
            actionStack,
            team,
            pulling,
        )
    }
}

export class TeamActionListData implements ActionListData {
    actionList: Action[]
    constructor(actionStack: LiveServerAction[], team: TeamNumber) {
        this.actionList = getTeamActionList(actionStack, team)
    }
}

const getPlayerActionList = (
    playerOne: DisplayUser,
    actionStack: LiveServerAction[],
    playerTeam: TeamNumber,
    pulling: boolean,
): Action[] => {
    let currentUser: string | undefined = playerOne._id
    for (const action of actionStack.slice().reverse()) {
        const playerTwo = action.playerOne
        switch (action.actionType) {
            case ActionType.PICKUP:
            case ActionType.CATCH:
                if (currentUser === action.playerOne?._id) {
                    return [new ThrowawayAction(playerOne)]
                } else {
                    return [
                        new CatchAction(playerOne, playerTwo),
                        new DropAction(playerOne, playerTwo),
                        new ScoreAction(
                            playerTeam,
                            'score',
                            playerOne,
                            playerTwo,
                        ),
                    ]
                }
            case ActionType.DROP:
            case ActionType.THROWAWAY:
            case ActionType.PULL:
                return [
                    new BlockAction(playerOne),
                    new PickupAction(playerOne),
                    new ScoreAction(playerTeam, 'clhn', playerOne, undefined),
                ]
            case ActionType.BLOCK:
                return [new PickupAction(playerOne)]
            case ActionType.TEAM_ONE_SCORE:
            case ActionType.TEAM_TWO_SCORE:
                return []
            case ActionType.TIMEOUT:
            case ActionType.CALL_ON_FIELD:
                continue
            case ActionType.SUBSTITUTION:
                // if we are getting actions for newly substituted user
                if (playerOne._id === action.playerTwo?._id) {
                    // inherit the actions from the player that was substituted for
                    currentUser = action.playerOne?._id
                }
                continue
        }
    }

    if (pulling) {
        return [new PullAction(playerOne)]
    }
    // first action on receiving team
    return [
        new CatchAction(playerOne),
        new PickupAction(playerOne),
        new DropAction(playerOne),
    ]
}

const getTeamActionList = (
    actionStack: LiveServerAction[],
    team: TeamNumber,
): Action[] => {
    for (const action of actionStack.slice().reverse()) {
        if (isRetainingPossession(action.actionType)) {
            return [
                new TimeoutAction(),
                new CallOnFieldAction(),
                new SubstitutionAction(),
            ]
        } else if (isTurnover(action.actionType)) {
            return [
                new ScoreAction(team === 'one' ? 'two' : 'one', 'they score'),
                new CallOnFieldAction(),
                new SubstitutionAction(),
            ]
        } else if (isScore(action.actionType)) {
            return []
        }
    }
    return []
}

class BaseAction implements Action {
    action: ServerAction
    reporterDisplay: string
    viewerDisplay: string

    constructor(
        action: Partial<ServerAction>,
        viewerDisplay: string,
        reporterDisplay?: string,
    ) {
        this.action = {
            actionNumber: Infinity,
            actionType: ActionType.PULL,
            tags: [],
            comments: [],
            ...action,
        }
        this.reporterDisplay = reporterDisplay ?? this.action.actionType
        this.viewerDisplay = viewerDisplay
    }

    setTags(tags: string[]): void {
        this.action.tags = tags
    }

    setPlayersAndUpdateViewerDisplay(
        playerOne?: DisplayUser,
        playerTwo?: DisplayUser,
    ): void {
        this.action.playerOne = playerOne
        this.action.playerTwo = playerTwo
    }

    setPlayers(playerOne?: DisplayUser, playerTwo?: DisplayUser) {
        this.action.playerOne = playerOne
        this.action.playerTwo = playerTwo
    }

    createFromAction(action: Partial<ServerAction>): Action {
        this.action = { ...this.action, ...action }
        return this
    }
}

class BlockAction extends BaseAction {
    constructor(playerOne: DisplayUser) {
        const playerOneDisplay = getUserDisplayName(playerOne)
        const viewerDisplay = `${playerOneDisplay} blocks the throw`
        super(
            {
                playerOne,
                actionType: ActionType.BLOCK,
            },
            viewerDisplay,
        )
    }
}

class CallOnFieldAction extends BaseAction {
    constructor() {
        super(
            { actionType: ActionType.CALL_ON_FIELD },
            'There is a call on the field',
            'call on field',
        )
    }
}

class CatchAction extends BaseAction {
    constructor(playerOne: DisplayUser, playerTwo?: DisplayUser) {
        super({ actionType: ActionType.CATCH }, 'Catch')
        this.setPlayersAndUpdateViewerDisplay(playerOne, playerTwo)
    }

    setPlayersAndUpdateViewerDisplay(
        playerOne?: DisplayUser,
        playerTwo?: DisplayUser,
    ) {
        this.setPlayers(playerOne, playerTwo)
        this.viewerDisplay = this.getViewerDisplay()
    }

    getViewerDisplay(): string {
        const playerOneDisplay = getUserDisplayName(this.action.playerOne)
        let viewerDisplay = `${playerOneDisplay} catches the disc`
        if (this.action.playerTwo) {
            const playerTwoDisplay = getUserDisplayName(this.action.playerTwo)
            viewerDisplay = viewerDisplay.concat(` from ${playerTwoDisplay}`)
        }
        return viewerDisplay
    }
}

class DropAction extends BaseAction {
    constructor(playerOne: DisplayUser, playerTwo?: DisplayUser) {
        super({ actionType: ActionType.DROP }, 'Drop')
        this.setPlayersAndUpdateViewerDisplay(playerOne, playerTwo)
    }

    setPlayersAndUpdateViewerDisplay(
        playerOne?: DisplayUser,
        playerTwo?: DisplayUser,
    ) {
        this.setPlayers(playerOne, playerTwo)
        this.viewerDisplay = this.getViewerDisplay()
    }

    getViewerDisplay(): string {
        const playerOneDisplay = getUserDisplayName(this.action.playerOne)
        let viewerDisplay = `${playerOneDisplay} drops the pass`
        if (this.action.playerTwo) {
            const playerTwoDisplay = getUserDisplayName(this.action.playerTwo)
            viewerDisplay = viewerDisplay.concat(` from ${playerTwoDisplay}`)
        }
        return viewerDisplay
    }
}

class PickupAction extends BaseAction {
    constructor(playerOne: DisplayUser) {
        const playerOneDisplay = getUserDisplayName(playerOne)
        const viewerDisplay = `${playerOneDisplay} picks up the disc`
        super({ playerOne, actionType: ActionType.PICKUP }, viewerDisplay)
    }
}

class PullAction extends BaseAction {
    constructor(playerOne: DisplayUser) {
        const playerOneDisplay = getUserDisplayName(playerOne)
        const viewerDisplay = `${playerOneDisplay} pulls the disc`
        super({ playerOne, actionType: ActionType.PULL }, viewerDisplay)
    }
}

class SubstitutionAction extends BaseAction {
    constructor(playerOne?: DisplayUser, playerTwo?: DisplayUser) {
        super({ actionType: ActionType.SUBSTITUTION }, 'substitution')
        this.setPlayersAndUpdateViewerDisplay(playerOne, playerTwo)
    }

    setPlayersAndUpdateViewerDisplay(
        playerOne?: DisplayUser,
        playerTwo?: DisplayUser,
    ) {
        this.action.playerOne = playerOne
        this.action.playerTwo = playerTwo
        if (playerOne && playerTwo) {
            const playerOneDisplay = getUserDisplayName(playerOne)
            const playerTwoDisplay = getUserDisplayName(playerTwo)
            this.viewerDisplay = `${playerTwoDisplay} replaces ${playerOneDisplay}`
        }
    }
}

class ScoreAction extends BaseAction {
    constructor(
        team: TeamNumber,
        reporterDisplay?: string,
        playerOne?: DisplayUser,
        playerTwo?: DisplayUser,
    ) {
        const actionType =
            team === 'one'
                ? ActionType.TEAM_ONE_SCORE
                : ActionType.TEAM_TWO_SCORE
        super({ actionType }, 'score', reporterDisplay)
        this.setPlayersAndUpdateViewerDisplay(playerOne, playerTwo)
    }

    setPlayersAndUpdateViewerDisplay(
        playerOne?: DisplayUser,
        playerTwo?: DisplayUser,
    ) {
        this.setPlayers(playerOne, playerTwo)
        this.viewerDisplay = this.getViewerDisplay()
    }

    getViewerDisplay() {
        let viewerDisplay = ''
        const playerOneDisplay = getUserDisplayName(this.action.playerOne)
        const playerTwoDisplay = getUserDisplayName(this.action.playerTwo)
        console.log(
            'score action got displays',
            this.action.playerOne,
            this.action.playerTwo,
        )
        if (playerOneDisplay && playerTwoDisplay) {
            viewerDisplay = `${playerOneDisplay} scores from ${playerTwoDisplay}`
        } else if (playerOneDisplay) {
            // TODO: is callahan the only way to get here?
            viewerDisplay = `${playerOneDisplay} scores a callahan`
        } else {
            viewerDisplay = `The opposing team scores`
        }
        return viewerDisplay
    }
}

class ThrowawayAction extends BaseAction {
    constructor(playerOne: DisplayUser) {
        const playerOneDisplay = getUserDisplayName(playerOne)
        const viewerDisplay = `${playerOneDisplay} throws the disc away`
        super({ playerOne, actionType: ActionType.THROWAWAY }, viewerDisplay)
    }
}

class TimeoutAction extends BaseAction {
    constructor() {
        super({ actionType: ActionType.TIMEOUT }, 'Timeout called')
    }
}

export class ActionFactory {
    static createFromAction = (action: ServerAction): Action => {
        switch (action.actionType) {
            case ActionType.BLOCK:
                return new BlockAction(action.playerOne!).createFromAction(
                    action,
                )
            case ActionType.CALL_ON_FIELD:
                return new CallOnFieldAction().createFromAction(action)
            case ActionType.CATCH:
                return new CatchAction(
                    action.playerOne!,
                    action.playerTwo,
                ).createFromAction(action)
            case ActionType.DROP:
                return new DropAction(
                    action.playerOne!,
                    action.playerTwo,
                ).createFromAction(action)
            case ActionType.PICKUP:
                return new PickupAction(action.playerOne!).createFromAction(
                    action,
                )
            case ActionType.PULL:
                return new PullAction(action.playerOne!).createFromAction(
                    action,
                )
            case ActionType.SUBSTITUTION:
                return new SubstitutionAction(
                    action.playerOne,
                    action.playerTwo,
                ).createFromAction(action)
            case ActionType.TEAM_ONE_SCORE:
            case ActionType.TEAM_TWO_SCORE:
                const scoreAction = new ScoreAction('one').createFromAction(
                    action,
                )
                scoreAction.setPlayersAndUpdateViewerDisplay(
                    action.playerOne,
                    action.playerTwo,
                )
                return scoreAction
            case ActionType.THROWAWAY:
                return new ThrowawayAction(action.playerOne!).createFromAction(
                    action,
                )
            case ActionType.TIMEOUT:
                return new TimeoutAction().createFromAction(action)
        }
    }
}
