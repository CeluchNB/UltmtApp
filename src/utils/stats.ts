import { DisplayStat, PlayerStats } from '../types/stats'

export const convertProfileScreenStatsToStatListItem = (
    stats?: PlayerStats,
): DisplayStat[] => {
    if (!stats) return []
    return [
        { name: 'Goals', value: stats.goals, points: stats.pointsPlayed },
        {
            name: 'Assists',
            value: stats.assists,
            points: stats.pointsPlayed,
        },
        { name: 'Blocks', value: stats.blocks, points: stats.pointsPlayed },
        {
            name: '+ / -',
            value: stats.plusMinus,
            points: stats.pointsPlayed,
        },
    ]
}

export const formatNumber = (key: string, value: number | string): string => {
    if (statIsPercentage(key)) {
        return `${(Number(value) * 100).toFixed(0)}%`
    }
    if (Number(value) % 1 === 0) {
        return value.toString()
    }
    return Number(value).toFixed(2)
}

const statIsPercentage = (key: string): boolean => {
    return key.toLowerCase().includes('percentage')
}

export const mapStatDisplayName = (value: string): string => {
    switch (value) {
        case 'completedPasses':
            return 'Completed Passes'
        case 'droppedPasses':
            return 'Dropped Passes'
        case 'pointsPlayed':
            return 'Points Played'
        case 'plusMinus':
            return 'Plus / Minus'
        case 'catchingPercentage':
            return 'Catching Percentage'
        case 'throwingPercentage':
            return 'Throwing Percentage'
        case 'ppGoals':
            return 'PP Goals'
        case 'ppAssists':
            return 'PP Assists'
        case 'ppThrowaways':
            return 'PP Throwaways'
        case 'ppDrops':
            return 'PP Drops'
        case 'ppBlocks':
            return 'PP Blocks'
        case 'winPercentage':
            return 'Win Percentage'
        default:
            return value.charAt(0).toUpperCase() + value.slice(1)
    }
}
