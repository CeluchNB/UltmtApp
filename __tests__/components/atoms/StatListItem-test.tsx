import * as React from 'react'
import { DisplayStat } from '../../../src/types/stats'
import StatListItem from '../../../src/components/atoms/StatListItem'
import { render } from '@testing-library/react-native'
import renderer from 'react-test-renderer'

const stat: DisplayStat = {
    name: 'Goals',
    value: 10,
    points: 100,
}

it('test matches snapshot', () => {
    const snapshot = renderer.create(<StatListItem stat={stat} />)
    expect(snapshot).toMatchSnapshot()
})

it('test renders correct view', () => {
    const { getByText } = render(<StatListItem stat={stat} />)

    const title = getByText('Goals')
    const value = getByText(stat.value.toString())

    expect(title).toBeTruthy()
    expect(value).toBeTruthy()
})
