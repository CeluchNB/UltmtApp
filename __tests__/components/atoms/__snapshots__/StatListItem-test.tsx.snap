// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test matches snapshot 1`] = `
<View
  style={
    {
      "minWidth": 150,
    }
  }
>
  <Text
    style={
      {
        "color": "#CEC6B4",
        "fontSize": 41.0958904109589,
        "fontWeight": "bold",
      }
    }
  >
    Goals
  </Text>
  <Text
    style={
      {
        "color": "#ffd600",
        "fontSize": 61.64383561643835,
        "fontWeight": "bold",
      }
    }
  >
    10
  </Text>
  <View
    style={
      {
        "backgroundColor": "#7AC6FF",
        "borderRadius": 5,
        "height": 10,
        "width": 10,
      }
    }
  />
</View>
`;
